src/diffusers/textual_inversion_utils.py

```diff
@@ -0,0 +1,101 @@
+        r"""
+        Loads textual inversion embeddings.
+
+        Receives a dictionary with the following keys:
+        - `token`: name of the token to be added to the tokenizers' vocabulary
+        - `embedding`: path to the embedding of the token to be added to the text encoder's embedding matrix
+
+        Alternatively, it can receive a list of pathes to embedding dictionaries, where the keys are the tokens and the
+        values are the embeddings. In that case, it will iterate through the list and add the tokens and embeddings to
+        the tokenizer's vocabulary and the text encoder's embedding matrix.
+
+        Iters through the dictionary and adds the token to the tokenizer's vocabulary and the embedding to the text
+        encoder's embedding matrix.
+        """
+
+        if isinstance(embeddings, dict):
+            for token, embedding_path in embeddings.items():
+                # check if token in tokenizer vocab
+                # if yes, raise exception
+                if token in self.tokenizer.get_vocab():
+                    raise ValueError(
+                        f"Token {token} already in tokenizer vocabulary. Please choose a different token name."
+                    )
+
+                embedding_dict = torch.load(embedding_path)
+                embedding = self.extract_embedding_from_dict(embedding_dict)
+
+                self.add_textual_inversion_embedding(token, embedding)
+
+        elif isinstance(embeddings, list):
+            for embedding_path in embeddings:
+                embedding_dict = torch.load(embedding_path)
+                token = self.extract_token_from_dict(embedding_dict)
+                embedding = self.extract_embedding_from_dict(embedding_dict)
+
+                # check if token in tokenizer vocab
+                # if yes, raise exception
+                if token in self.tokenizer.get_vocab():
+                    raise ValueError(
+                        f"Token {token} already in tokenizer vocabulary. Please choose a different token name."
+                    )
+                self.add_textual_inversion_embedding(token, embedding)
+
+    def extract_embedding_from_dict(self, embedding_dict):
+        r"""
+        Extracts the embedding from the embedding dictionary.
+        """
+        # auto1111 embedding case
+        if "string_to_param" in embedding_dict:
+            embedding_dict = embedding_dict["string_to_param"]
+            embedding = embedding_dict["*"]
+            return embedding
+
+        return list(embedding_dict.values())[0]
+
+    def extract_token_from_dict(self, embedding_dict):
+        r"""
+        Extracts the token from the embedding dictionary.
+        """
+        # auto1111 embedding case
+        if "string_to_param" in embedding_dict:
+            token = embedding_dict["name"]
+            return token
+
+        return list(embedding_dict.keys())[0]
+
+    def add_textual_inversion_embedding(self, token, embedding):
```

Comment:
Can we maybe make this function a bit more general to also load multiple tokens & embeddings ? Also cc @patil-suraj

Reply:
It does it on other method. Unless you mean multiple tokens and embeddings on the same file. Is that the case?

Reply:
Support for a list of files could be useful. In my code I test the filename and if it is a file I set `files=[filename]` and otherwise `files=glob.glob(os.path.join(filename, "*"))`. Afterward I can iterate over `files` and load all embeddings.
The advantage of having the loader handle multiple files is, that you only resize once and `tokenizer` can add a list of tokens at the same time.

Loading multiple tokens from each file is interesting as well. The two features are not mutually exclusive.

Reply:
> Support for a list of files could be useful. In my code I test the filename and if it is a file I set `files=[filename]` and otherwise `files=glob.glob(os.path.join(filename, "*"))`. Afterward I can iterate over `files` and load all embeddings. The advantage of having the loader handle multiple files is, that you only resize once and `tokenizer` can add a list of tokens at the same time.

It should support the use case of passing in a list of files; however, the implementation loads the embeddings sequentially and resizes sequentially. This can be improved.

> Loading multiple tokens from each file is interesting as well. The two features are not mutually exclusive.
In the case of Auto1111, a user can define a higher dimension embedding for a token and I believe this is where a lot of people in the community run into issues. So, for example, a mapping may be `token --> Embedding.shape = (6, 768)`.

I'm currently trying to implement the same ability in a HF diffusers compatible textual inversion training script and have traced down how the higher dimensional embeddings are instantiated, passed to the optimizer, and then later loaded. It's late now but I'll share the HF and Auto1111 code snippets tomorrow juxtaposed.

Hopefully, you can shed some additional light on what changes need to be made to achieve compatibility.

Reply:
Referencing PR #661 here because they are related and should be compatible.